conu/backend/nspawn/image.py

```diff
@@ -0,0 +1,332 @@
+                return True
+        raise ConuException("Unable to stop machine %s within %d" % (name, constants.DEFAULT_RETRYTIMEOUT))
+
+    def _is_booted(self, name, suffictinet_texts=["systemd-logind"]):
+        # optionally use: "Unit: machine"
+        for foo in range(constants.DEFAULT_RETRYTIMEOUT):
+            time.sleep(constants.DEFAULT_SLEEP)
+            out = run_cmd(["machinectl", "status", name], ignore_status=True, return_output=True)
+            for restr in suffictinet_texts:
+                if restr in out:
+                    time.sleep(constants.DEFAULT_SLEEP)
+                    return True
+        raise ConuException("Unable to start machine %s within %d" % (name, constants.DEFAULT_RETRYTIMEOUT))
+
+    def run(self, command=[], volumes=[], additional_opts=[], default_options=["-b"], foreground=False, name=None, *args, **kwargs):
+        logger.info("run container via binary in background")
+        machine_name = constants.CONU_ARTIFACT_TAG
+        if name:
+            machine_name += name
+        else:
+            machine_name += random_str()
+        if not kwargs:
+            kwargs = {}
+        if not foreground:
+            kwargs["stdout"] = subprocess.PIPE
+        additional_opts += default_options
+        if volumes:
+            additional_opts += self.get_volume_options(volumes=volumes)
+        logger.debug("starting NSPAWN")
+        systemd_command = ["systemd-nspawn", "--machine", machine_name, "-i", self.get_metadata()["Path"]] + additional_opts + command
+        logger.debug("Start command: %s" % " ".join(systemd_command))
+        self.container_process = subprocess.Popen(systemd_command, *args, **kwargs)
+        logger.info("machine: %s starting" % self.name)
+        if foreground:
+            self.container_process.wait()
+            return self.container_process
+        else:
+            self._is_booted(machine_name)
+        logger.info("machine: %s starting finished" % machine_name)
+        return NspawnContainer(self, container_id=machine_name, )
+
+    def run_foreground(self, *args, **kwargs):
+        return self.run(foreground=True, default_options=[], *args, **kwargs)
+
+
+    @staticmethod
+    def get_volume_options(volumes):
+        """
+        Generates volume options to run methods.
+
+        :param volumes: tuple or list of tuples in form target x source,target x source,target,mode.
+        :return: list of the form ["--bind", "/source:/target", "--bind", "/other/source:/destination:z", ...]
+        """
+        if not isinstance(volumes, list):
+            volumes = [volumes]
+        volumes = [Volume.create_from_tuple(v) for v in volumes]
+        result = []
+        for v in volumes:
+            result += ["--bind", str(v)]
+        return result
+
+    @staticmethod
+    def bootstrap(repositories, packages=constants.CONU_NSPAWN_BASEPACKAGES, additional_packages=[], tag="latest", prefix=constants.CONU_ARTIFACT_TAG,
```

Comment:
also please fix the default values: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments